src/MediaWiki/Search/Search.php

```diff
@@ -163,6 +104,52 @@ private function searchFallbackSearchEngine( $term, $fulltext ) {
 	}
 
 	/**
+	 * @return SearchEngine
+	 */
+	public function getFallbackSearchEngine() {
+		if ( $this->fallbackSearch === null ) {
+			$type = ApplicationFactory::getInstance()->getSettings()->get( 'smwgFallbackSearchType' );
+
+			$dbr = $this->getDB();
+
+			if ( $type === null ) {
+				$type = $this->getDefaultSearchEngineTypeForDB( $dbr );
+			}
+
+			$this->assertIsSearchEngineClassName( $type );
+
+			$this->fallbackSearch = new $type( $dbr );
+		}
+
+		return $this->fallbackSearch;
+	}
+
+	/**
+	 * @param DatabaseBase $db
+	 * @return string
+	 */
+	private function getDefaultSearchEngineTypeForDB( DatabaseBase $db ) {
+		if ( method_exists( 'SearchEngineFactory', 'getSearchEngineClass' ) ) { // MW > 1.27
+			return \SearchEngineFactory::getSearchEngineClass( $db );
+		}
+
+		return $db->getSearchEngine();
+	}
+
+	/**
+	 * @param string $className
+	 */
+	private function assertIsSearchEngineClassName( $className ) {
+		if ( !class_exists( $className ) ) {
+			throw new RuntimeException( "$className does not exist." );
+		}
+
+		if ( !is_subclass_of( $className, 'SearchEngine' ) || $className === 'SMWSearch' ) {
```

Comment:
@s7eph4n This predicate is as you wrote it, and I'm not sure it's correct. We should not get an exception when `$className === 'SMWSearch'` right? So some brackets are missing... though even in that case, I'm not sure why you would have this, since `SMWSearch` is a subclass of `SearchEngine` no?

I should not have merged your PR before that is addressed though saw too late.